diff --git a/utils/lit/lit/TestRunner.py b/utils/lit/lit/TestRunner.py
index 9752417..e5eb9b4 100644
--- a/utils/lit/lit/TestRunner.py
+++ b/utils/lit/lit/TestRunner.py
@@ -296,8 +296,12 @@ def executeScript(test, litConfig, tmpBase, commands, cwd):
             # run on clang with no real loss.
             command = litConfig.valgrindArgs + command
 
+    dredd_env=test.config.environment.copy()
+    if 'DREDD_ENABLED_MUTATION' in os.environ:
+        dredd_env['DREDD_ENABLED_MUTATION'] = os.environ['DREDD_ENABLED_MUTATION']
+
     return lit.util.executeCommand(command, cwd=cwd,
-                                   env=test.config.environment)
+                                   env=dredd_env)
 
 def parseIntegratedTestScriptCommands(source_path):
     """
