diff --git a/utils/lit/lit/TestRunner.py b/utils/lit/lit/TestRunner.py
index a7de2b7..a7303e1 100644
--- a/utils/lit/lit/TestRunner.py
+++ b/utils/lit/lit/TestRunner.py
@@ -274,8 +274,12 @@ def executeTclScriptInternal(test, litConfig, tmpBase, commands, cwd):
             return '', '', 0
 
         command = [litConfig.getBashPath(), script]
+
+        dredd_env=test.config.environment.copy()
+        if 'DREDD_MUTANT_TRACKING_FILE' in os.environ:
+            dredd_env['DREDD_MUTANT_TRACKING_FILE'] = os.environ['DREDD_MUTANT_TRACKING_FILE']
         out,err,exitCode = executeCommand(command, cwd=cwd,
-                                          env=test.config.environment)
+                                          env=dredd_env)
 
         # Tcl commands fail on standard error output.
         if err:
@@ -334,7 +338,11 @@ def executeScript(test, litConfig, tmpBase, commands, cwd):
 
             command = valgrindArgs + command
 
-    return executeCommand(command, cwd=cwd, env=test.config.environment)
+    dredd_env=test.config.environment.copy()
+    if 'DREDD_MUTANT_TRACKING_FILE' in os.environ:
+        dredd_env['DREDD_MUTANT_TRACKING_FILE'] = os.environ['DREDD_MUTANT_TRACKING_FILE']
+
+    return executeCommand(command, cwd=cwd, env=dredd_env)
 
 def isExpectedFail(xfails, xtargets, target_triple):
     # Check if any xfail matches this target.
